gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -0,0 +1,281 @@
+      }
+    }
+  }
+
+  private KafkaPartition getKafkaPartitionFromWorkUnitState(WorkUnitState workUnitState) {
+    if (workUnitState.contains(TOPIC_NAME) && workUnitState.contains(PARTITION_ID)) {
+      return new KafkaPartition.Builder().withTopicName(workUnitState.getProp(TOPIC_NAME))
+          .withId(workUnitState.getPropAsInt(PARTITION_ID)).build();
+    }
+    return null;
+  }
+
+  /**
+   * A topic can be configured to move to the latest offset in topics.move.to.latest.offset
+   */
+  private boolean shouldMoveToLatestOffset(KafkaPartition partition, SourceState state) {
+    if (!state.contains(TOPICS_MOVE_TO_LATEST_OFFSET)) {
+      return false;
+    }
+    if (this.moveToLatestTopics.isEmpty()) {
+      this.moveToLatestTopics.addAll(Splitter.on(',').trimResults().omitEmptyStrings()
+          .splitToList(state.getProp(TOPICS_MOVE_TO_LATEST_OFFSET)));
+    }
+    return this.moveToLatestTopics.contains(partition.getTopicName()) || moveToLatestTopics.contains("all");
+  }
+
+  private WorkUnit getWorkUnitForTopicPartition(KafkaPartition partition, SourceState state, Offsets offsets) {
+    SourceState partitionState = createAndPopulatePartitionState(partition, state, offsets);
+    Extract extract = createExtract(partitionState, partition.getTopicName());
+    LOG.info(String.format("Creating workunit for topic %s, partition %d: lowWatermark=%d, highWatermark=%d",
+        partition.getTopicName(), partition.getId(), offsets.getStartOffset(), offsets.getLatestOffset()));
+    return partitionState.createWorkUnit(extract);
+  }
+
+  private Extract createExtract(SourceState partitionState, String topicName) {
+    return partitionState.createExtract(DEFAULT_TABLE_TYPE, DEFAULT_NAMESPACE_NAME, topicName);
+  }
+
+  private SourceState createAndPopulatePartitionState(KafkaPartition partition, SourceState state, Offsets offsets) {
+    SourceState partitionState = new SourceState();
+    partitionState.addAll(state);
+    partitionState.setProp(TOPIC_NAME, partition.getTopicName());
+    partitionState.setProp(PARTITION_ID, partition.getId());
+    partitionState.setProp(LEADER_ID, partition.getLeader().getId());
+    partitionState.setProp(LEADER_HOST, partition.getLeader().getHost());
+    partitionState.setProp(LEADER_PORT, partition.getLeader().getPort());
+    partitionState.setProp(ConfigurationKeys.WORK_UNIT_LOW_WATER_MARK_KEY, offsets.getStartOffset());
+    partitionState.setProp(ConfigurationKeys.WORK_UNIT_HIGH_WATER_MARK_KEY, offsets.getLatestOffset());
+    return partitionState;
+  }
+
+  private boolean movingToEarliestOffsetAllowed(SourceState state) {
+    return state.getPropAsBoolean(MOVE_TO_EARLIEST_OFFSET_ALLOWED, DEFAULT_MOVE_TO_EARLIEST_OFFSET_ALLOWED);
+  }
+
+  private List<KafkaTopic> getFilteredTopics(KafkaWrapper kafkaWrapper, SourceState state) {
+    Set<String> blacklist =
+        state.contains(TOPIC_BLACKLIST) ? state.getPropAsCaseInsensitiveSet(TOPIC_BLACKLIST) : new HashSet<String>();
```

Comment:
Please be consistent in the ways you create new Lists, Sets, or Maps. Since you use Lists, Sets, Maps in other places, please also use them here.